src/Common/FileCache.cpp

```diff
@@ -480,8 +486,185 @@ FileSegmentsHolder LRUFileCache::setDownloading(const Key & key, size_t offset,
     return FileSegmentsHolder(std::move(file_segments));
 }
 
-bool LRUFileCache::tryReserve(
-    const Key & key, size_t offset, size_t size, std::lock_guard<std::mutex> & cache_lock)
+LRUFileCache::QueryContextPtr LRUFileCache::getCurrentQueryContext(std::lock_guard<std::mutex> &) const
+{
+    if (!isQueryInitialized())
+        return nullptr;
+
+    auto query_iter = query_map.find(CurrentThread::getQueryId().toString());
+    return (query_iter == query_map.end()) ? nullptr : query_iter->second;
+}
+
+LRUFileCache::QueryContextPtr LRUFileCache::getQueryContext(const String & query_id, std::lock_guard<std::mutex> &) const
+{
+    auto query_iter = query_map.find(query_id);
+    return (query_iter == query_map.end()) ? nullptr : query_iter->second;
+}
+
+bool LRUFileCache::tryReserve(const Key & key, size_t offset, size_t size, std::lock_guard<std::mutex> & cache_lock)
+{
+    auto query_context = getCurrentQueryContext(cache_lock);
+
+    /// If the context can be found, subsequent cache replacements are made through the Query context.
+    if (query_context != nullptr)
```

Comment:
this also needs to be added under setting like `filesystem_cache_use_query_cache_limit`, by default `false`, so by default no query context is saved